### ISSUE TITLE
OSSL_DECODER_from_bio fails but reports no error

### ISSUE THREAD

**Issue by Jezza2** [openssl/openssl#21736]
Steps to reproduce:

```Console
$ openssl genpkey -algorithm EC -out ec.pem -pkeyopt ec_param_gen_curve:sect239k1
```

```C
#include <openssl/err.h>
#include <openssl/decoder.h>
#include <openssl/bio.h>
#include <openssl/evp.h>
#include <stdio.h>

int main()
{
    BIO* b = BIO_new_file("ec.pem", "rb");
    EVP_PKEY* pkey = NULL;
    OSSL_DECODER_CTX* ctx = OSSL_DECODER_CTX_new_for_pkey(
            &pkey, NULL, NULL, NULL, EVP_PKEY_PUBLIC_KEY, NULL, NULL);

    int r = OSSL_DECODER_from_bio(ctx, b);
    if (r == 1)
    {
        EVP_PKEY_print_params_fp(stdout, pkey, 0, NULL);
        return 0;
    }
    else
    {
        ERR_print_errors_fp(stdout);
        return 1;
    }

    return 0;
}
```

Running this code prints nothing (i.e. `OSSL_DECODER_from_bio()` failed, but there is no error on the stack). If I add the key type to PEM header (i.e. `----BEGIN PRIVATE KEY-----` becomes `----BEGIN EC PRIVATE KEY-----`) then an error is printed: `...:No supported data to decode`.

I've observed this behaviour with versions 3.0.10 and 3.1.2, both built from source on Debian 11 with default options.

**Comment by nhorman** [comment]
I'd check the return value of  OSSL_DECODER_CTX_new_for_pkey.  If that call allocated a new ctx, but failed to initalize it for some reason (like if ossl_decoder_ctx_setup_for_pkey failed), you would get a NULL ctx with nothing on the error stack.  We should likely raise an error there, but that would be my guess at the cause.

**Comment by nhorman** [comment]
so, I'm sorry, I ran your test case, and this  is a bug.  Augmenting your code, we definately get a ctx, and it lists 20 identified decoders that can address this input data, but none of them successfully decode the pem file

**Comment by Jezza2** [comment]
Yes - sorry I didn't notice your first reply until today. I was certain the ctx wasn't NULL where I was seeing this issue, so I didn't bother checking it in my test case.

Presumably the bug is twofold?
1. One of the decoders should be able to decode the pem file
2. If none of them can, an error should be reported.

I also wondered if the error 'No supported data to decode' is correct behaviour. After reading the [docs](https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_fromdata.html), I would expect the selection `EVP_PKEY_PUBLIC_KEY` to succeed when reading a private key file.

**Comment by nhorman** [comment]
one of the decoders should certainly be able to decode the file yes.

The error report issue seems to have been fixed.  When I generated my pem file here under the latest build and attempted to decode it, I immediately got the no decoders found error with the unmodified pem file

**Comment by t8m** [comment]
The decoders handle the selection quite differently from the fromdata call. And I am afraid changing this will have bad implications. Like that after the change decoding for a PUBLIC_KEY will start asking for a password even if it would not before.

Maybe we should just document this.

You can also use 0 as the selection and that should allow decoding the public key, but of course the resulting EVP_PKEY will have both public and private key parts if the decoded file contained it.

**Comment by t8m** [comment]
@levitte what do you think about this?

**Comment by nhorman** [comment]
I'm sorry I missed what you said above @Jezza2:
openssl genpkey -algorithm EC -out ec.pem -pkeyopt ec_param_gen_curve:sect239k1
generates a private ec key
but in your sample code you're selecting EVP_PKEY_PUBLIC_KEY

No, that won't read a private key because the data types won't match.  the pem2der_decode function correctly identifies your private key as a private key data type, and when we attempt to convert from der to an EVP_PKEY, your selection of EVP_PUBLIC_KEY doesn't match the data type, so it doesn't import it.

If I change EVP_PKEY_PUBLIC_KEY to  OSSL_KEYMGMT_SELECT_PRIVATE_KEY in your selection, the decode works just fine (though we should update the documentation to reflect the selection values properly).

This seems to be working like it should however.

I'll get to fixing up the documentation shortly

**Comment by levitte** [comment]
From a PEM and DER perspective, a public key is a different object than a private key.  They are different structures.  Therefore, when you ask for a public key, it can be argued that if the file as a whole is a private key, then the ask doesn't match the reality of that file.

This, BTW, matches how things were done in OpenSSL before 3.0.  You can't expect to get a public key with for example `PEM_read_bio_PUBKEY()` when presented with this same file.

It is of course debatable that *the contents* of a private key usually has all the bits necessary for a public key as well.  For example, with an RSA key, the items *n* and *e* are present in the private key data as well (this is defined by the PKCS#1 spec)...
However, specifically for the EC key on file, *it really only contains the private key*, so even so, one can't really expect to get a public key out of it.
(and mind you, this has been a subject for debate)

Does this help?

**Comment by nhorman** [comment]
Marking as inactive due to lack of response.  Will be closed at the end of the 3.4 dev cycle

**Comment by OrangeDog** [comment]
What was the resolution here? Do you now get an error set when the decode fails? Was/will that be backported to the 3.0 etc. branches?
